resilience4j-bulkhead/src/main/java/io/github/resilience4j/bulkhead/adaptive/internal/AdaptiveLimitBulkhead.java

```diff
@@ -0,0 +1,226 @@
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *
+ */
+package io.github.resilience4j.bulkhead.adaptive.internal;
+
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import io.github.resilience4j.bulkhead.Bulkhead;
+import io.github.resilience4j.bulkhead.BulkheadConfig;
+import io.github.resilience4j.bulkhead.adaptive.AdaptiveBulkhead;
+import io.github.resilience4j.bulkhead.adaptive.AdaptiveBulkheadConfig;
+import io.github.resilience4j.bulkhead.adaptive.LimitPolicy;
+import io.github.resilience4j.bulkhead.adaptive.internal.amid.AIMDLimiter;
+import io.github.resilience4j.bulkhead.adaptive.internal.config.AIMDConfig;
+import io.github.resilience4j.bulkhead.event.BulkheadLimit;
+import io.github.resilience4j.bulkhead.event.BulkheadOnLimitDecreasedEvent;
+import io.github.resilience4j.bulkhead.event.BulkheadOnLimitIncreasedEvent;
+import io.github.resilience4j.bulkhead.internal.SemaphoreBulkhead;
+import io.github.resilience4j.core.EventConsumer;
+import io.github.resilience4j.core.EventProcessor;
+import io.github.resilience4j.core.EventPublisher;
+import io.github.resilience4j.core.lang.NonNull;
+import io.github.resilience4j.core.lang.Nullable;
+
+public class AdaptiveLimitBulkhead implements AdaptiveBulkhead {
+	private final static AdaptiveBulkheadEventProcessor eventProcessor = new AdaptiveBulkheadEventProcessor();
+	private final String name;
+	private final AdaptiveBulkheadConfig adaptationConfig;
+	private final InternalMetrics metrics;
+	private final SemaphoreBulkhead bulkhead;
+	private final AtomicInteger inFlight = new AtomicInteger();
+	// current settings and measurements that you can read concurrently to expose metrics
+	private final BulkheadConfig currentConfig; // immutable object
+	@NonNull
+	private final LimitPolicy<Bulkhead> limitAdapter;
+
+	private AdaptiveLimitBulkhead(@NonNull String name, @NonNull AdaptiveBulkheadConfig config, @NonNull LimitPolicy<Bulkhead> limitAdapter, BulkheadConfig bulkheadConfig) {
+		this.name = name;
+		this.limitAdapter = limitAdapter;
+		this.adaptationConfig = config;
+		this.currentConfig = bulkheadConfig;
+		bulkhead = new SemaphoreBulkhead(name + "-internal", this.currentConfig);
+		metrics = new InternalMetrics();
+	}
+
+	@Override
+	public boolean tryAcquirePermission() {
+		boolean isAcquire = bulkhead.tryAcquirePermission();
+		if (isAcquire) {
+			inFlight.incrementAndGet();
+		}
+		return isAcquire;
+	}
+
+	@Override
+	public void acquirePermission() {
+		bulkhead.acquirePermission();
+		inFlight.incrementAndGet();
+	}
+
+	@Override
+	public void releasePermission() {
+		bulkhead.releasePermission();
+		inFlight.decrementAndGet();
+	}
+
+	@Override
+	public void onComplete(Duration callTime, boolean isSuccess) {
```

Comment:
Can we replace `onComplete` by `onSuccess` and `onError` to make the API comparable to the CircuitBreaker.

Reply:
Please use `long duration, TimeUnit durationUnit` instead of `Duration callTime`.
The reasons is that we don't want to create a Duration object for every call. 
TimeUnit instead is just an enum and a primitve. It should result in less garbage.